File: DependencyProcessor.py
import SCC

class DependencyGroup(object):
    def __init__(self):
        self.items = []

def createIdMaps(graph, all_dependencies):
    id_item_map = {}
    item_id_map = {}
    for key in all_dependencies.keys():
         id = graph.addNode()
         id_item_map[id] = key
         item_id_map[key] = id
    return (item_id_map, id_item_map)

def initGraph(graph, item_id_map, all_dependencies):
     for (item, deps) in all_dependencies.items():
          item_id = item_id_map[item]
          for dep in deps:
               dep_id = item_id_map[dep]
               graph.addNeighbour(item_id, dep_id)

def processDependencies(all_dependencies):
    graph = SCC.Graph()
    (item_id_map, id_item_map) = createIdMaps(graph, all_dependencies)
    initGraph(graph, item_id_map, all_dependencies)
    sccs = graph.collectSCCs()
    ordered_groups = []
    for scc in sccs:
        items = []
        for i in scc:
            items.append(id_item_map[i])
        group = DependencyGroup()
        group.items = items
        ordered_groups.append(group)
    return ordered_groups

File: compiler/src/siko/util/DependencyProcessor.rs
use std::{
    collections::BTreeMap,
    fmt::{Debug, Display},
};

use super::SCC::{self, Graph};

#[derive(Clone)]
pub struct DependencyGroup<T> {
    pub items: Vec<T>,
}

impl<T: Debug> Debug for DependencyGroup<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self.items)
    }
}

fn createIdMaps<T: Ord + Clone>(
    graph: &mut Graph,
    all_dependencies: &BTreeMap<T, Vec<T>>,
) -> (BTreeMap<SCC::NodeId, T>, BTreeMap<T, SCC::NodeId>) {
    let mut id_item_map = BTreeMap::new();
    let mut item_id_map = BTreeMap::new();
    for key in all_dependencies.keys() {
        let id = graph.addNode();
        id_item_map.insert(id, key.clone());
        item_id_map.insert(key.clone(), id);
    }
    (id_item_map, item_id_map)
}

fn initGraph<T: Ord + Clone + Display + Debug>(
    graph: &mut Graph,
    item_id_map: &BTreeMap<T, SCC::NodeId>,
    all_dependencies: &BTreeMap<T, Vec<T>>,
) {
    for (item, deps) in all_dependencies {
        let item_id = item_id_map.get(item).unwrap();
        for dep in deps {
            if let Some(dep_id) = item_id_map.get(dep) {
                graph.addNeighbour(*item_id, *dep_id);
            } else {
                panic!("Dependency {:?} not found in item_id_map", dep);
            }
        }
    }
}

pub fn processDependencies<T: Ord + Clone + Display + Debug>(
    all_dependencies: &BTreeMap<T, Vec<T>>,
) -> Vec<DependencyGroup<T>> {
    let mut graph = Graph::new();
    let (id_item_map, item_id_map) = createIdMaps(&mut graph, &all_dependencies);
    initGraph(&mut graph, &item_id_map, &all_dependencies);
    let sccs = graph.collectSCCs();
    let mut ordered_groups = Vec::new();
    for scc in sccs {
        let mut items = Vec::new();
        for i in scc {
            items.push(id_item_map.get(&i).unwrap().clone());
        }
        let group = DependencyGroup { items: items };
        ordered_groups.push(group);
    }
    return ordered_groups;
}
